### PR TITLE
Convert core to TS

### PR DESCRIPTION
**Pull request by kristw**
🏠 Internal
- Convert all code in `@superset-ui/core` to TypeScript
- Add generic definition for `Registry`

### CODE REVIEW

File: packages/superset-ui-core/src/models/Registry.ts [hunk 1]
```diff
@@ -59,14 +80,14 @@ export default class Registry<V> {
 
   registerLoader(key: string, loader: () => V | Promise<V>) {
```
[the remark this answers is not in the document]
Comment (kristw) [reply packages/superset-ui-core/src/models/Registry.ts]: `ItemWithLoader` is the stored item in the lookup which is `{ loader }` while the parameter here is `loader`.

**Comment by kristw** [comment]
Run into issue with `eslint-plugin-typescript` `no-unused-vars` rule with generic.
Might be related to this issue: https://github.com/bradzacher/eslint-plugin-typescript/issues/179

File: packages/superset-ui-core/src/models/Preset.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
+import Plugin from './Plugin';
+
+export interface PresetConfig {
+  name?: string;
+  description?: string;
+  presets?: Preset[];
+  plugins?: Plugin[];
+}
+
+export class Preset {
```
Comment (williaster) [inline packages/superset-ui-core/src/models/Preset.ts]: I thought we decided on default exports for classes/components, etc. and named exports for types/interfaces?

File: packages/superset-ui-core/src/models/Registry.ts [hunk 3]
```diff
@@ -108,26 +144,26 @@ export default class Registry {
     );
   }
 
-  keys() {
+  keys(): string[] {
     return Object.keys(this.items);
   }
 
-  values() {
+  values(): Array<V | Promise<V> | undefined> {
```
Comment (williaster) [inline packages/superset-ui-core/src/models/Registry.ts]: looks like you can also do `(type | type | type)[]`, the `Array<..>` generic seems a little more readable for this case tho so 🤷‍♂️

Comment (kristw) [reply packages/superset-ui-core/src/models/Registry.ts]: I will keep the `Array<>` as it is more readable. Will use `[]` for primitive and types.

**Pull request merged**